File: packages/portico-client-core/tests/offlineDownloadAuthorization.test.mjs
import assert from "node:assert/strict";
import {createHash, createPublicKey, generateKeyPairSync, sign, verify} from "node:crypto";
import test from "node:test";

import {
  canonicalOfflineDownloadAuthorizationPayload,
  parseOfflineDownloadAuthorizationReceipt,
  pinnedServerIdentityFromDurableSource,
  revalidateOfflineDownloadAuthorization,
  validateOfflineDownloadAuthorizationReceipt
} from "../dist/offlineDownloadAuthorization.js";

const {privateKey, publicKey} = generateKeyPairSync("ed25519");
const publicDER = publicKey.export({format: "der", type: "spki"});
const rawPublicKey = new Uint8Array(publicDER.subarray(publicDER.length - 32));
const base64url = value => Buffer.from(value).toString("base64url");
const fingerprint = `sha256:${base64url(createHash("sha256").update(rawPublicKey).digest())}`;
const runtime = {
  encodeBase64: base64url,
  decodeBase64: value => new Uint8Array(Buffer.from(value, "base64")),
  encodeText: value => new TextEncoder().encode(value),
  sha256: value => new Uint8Array(createHash("sha256").update(value).digest()),
  verifyEd25519: ({publicKey: raw, signature, message}) => {
    const key = createPublicKey({key: Buffer.concat([Buffer.from("302a300506032b6570032100", "hex"), Buffer.from(raw)]), format: "der", type: "spki"});
    return verify(null, Buffer.from(message), key, Buffer.from(signature));
  }
};

const scopeA = Object.freeze({authority: "local", accountId: "account-a", profileId: "profile-a", serverId: "server-a", authorizationRevision: "revision-7"});
const scopeB = Object.freeze({...scopeA, profileId: "profile-b", authorizationRevision: "revision-9"});
const artifact = Object.freeze({sha256: createHash("sha256").update("protected-bytes").digest("hex"), sizeBytes: 15});
const pinnedIdentity = Object.freeze({serverId: "server-a", publicKeyFingerprint: fingerprint, publicKey: rawPublicKey});
const encodedPublicKey = Buffer.from(rawPublicKey).toString("base64").replace(/=+$/u, "");

function signedReceipt({receiptId = "receipt-1", viewerScope = scopeA, artifactOverride = artifact, lastVerifiedAt = "2026-08-30T12:00:00.000Z"} = {}) {
  const receipt = {
    version: 1,
    purpose: "offline-download-authorization",
    receiptId,
    viewerScope: {scopeKind: "server-bound", ...viewerScope},
    issuer: {serverId: viewerScope.serverId, signingKeyFingerprint: fingerprint},
    preparation: {preparationId: "preparation-1", mediaId: "media-1", mediaVersionId: "version-1", qualityId: "source"},
    artifact: artifactOverride,
    lastVerifiedAt,
    verifyBy: new Date(Date.parse(lastVerifiedAt) + 2_592_000_000).toISOString(),
    signature: ""
  };
  receipt.signature = sign(null, Buffer.from(canonicalOfflineDownloadAuthorizationPayload(receipt)), privateKey).toString("base64url");
  return Object.freeze(receipt);
}

function context(activeViewerScope = scopeA, overrides = {}) {
  return {
    binding: {
      storedViewerScope: scopeA,
      originatingServerId: "server-a",
      preparationId: "preparation-1",
      mediaId: "media-1",
      mediaVersionId: "version-1",
      qualityId: "source",
      artifactSha256: artifact.sha256,
      artifactSizeBytes: artifact.sizeBytes,
      ...overrides
    },
    activeViewerScope,
    pinnedIdentity,
    runtimeAdapters: runtime,
    now: Date.parse("2026-08-31T12:00:00Z")
  };
}

test("durable connection identity supplies the exact offline receipt pin", async () => {
  const derived = await pinnedServerIdentityFromDurableSource({
    serverId: "server-a",
    serverPublicKey: encodedPublicKey,
    serverPublicKeyFingerprint: fingerprint
  }, runtime);
  assert.equal(derived.serverId, "server-a");
  assert.equal(derived.publicKeyFingerprint, fingerprint);
  assert.deepEqual(derived.publicKey, rawPublicKey);
  await assert.rejects(pinnedServerIdentityFromDurableSource({
    serverId: "server-a",
    serverPublicKey: encodedPublicKey,
    serverPublicKeyFingerprint: `sha256:${"A".repeat(43)}`
  }, runtime), /does not match/);
  await assert.rejects(pinnedServerIdentityFromDurableSource({
    serverId: "server-a",
    serverPublicKey: "",
    serverPublicKeyFingerprint: fingerprint
  }, runtime), /incomplete/);
});

test("receipt JCS is member-order independent and preserves Unicode scalars", () => {
  const receipt = signedReceipt({receiptId: "réceipt-😀"});
  const reordered = {
    signature: receipt.signature, verifyBy: receipt.verifyBy, lastVerifiedAt: receipt.lastVerifiedAt,
    artifact: {sizeBytes: receipt.artifact.sizeBytes, sha256: receipt.artifact.sha256},
    preparation: {...receipt.preparation}, issuer: {...receipt.issuer}, viewerScope: {...receipt.viewerScope},
    receiptId: receipt.receiptId, purpose: receipt.purpose, version: receipt.version
  };
  assert.equal(canonicalOfflineDownloadAuthorizationPayload(reordered), canonicalOfflineDownloadAuthorizationPayload(receipt));
  assert.equal(parseOfflineDownloadAuthorizationReceipt(reordered).receiptId, "réceipt-😀");
  assert.throws(() => parseOfflineDownloadAuthorizationReceipt({...reordered, receiptId: "bad\ud800"}));
  assert.throws(() => parseOfflineDownloadAuthorizationReceipt({...reordered, signature: `${receipt.signature.slice(0, -1)}B`}));
});

test("valid receipt is strict-before verifyBy and wrong active profile is non-destructive", async () => {
  const receipt = signedReceipt();
  assert.equal((await validateOfflineDownloadAuthorizationReceipt(receipt, context())).state, "valid");
  const wrongActiveProfile = await validateOfflineDownloadAuthorizationReceipt(receipt, context(scopeB));
  assert.deepEqual(wrongActiveProfile, {state: "out-of-scope", deleteProtectedArtifact: false});
  const corruptStoredReceiptUnderOtherProfile = await validateOfflineDownloadAuthorizationReceipt({}, context(scopeB));
  assert.deepEqual(corruptStoredReceiptUnderOtherProfile, {state: "out-of-scope", deleteProtectedArtifact: false});
  const atBoundary = await validateOfflineDownloadAuthorizationReceipt(receipt, {...context(), now: Date.parse(receipt.verifyBy)});
  assert.equal(atBoundary.state, "authorization-unverified");
  const intrinsicDigestMismatch = await validateOfflineDownloadAuthorizationReceipt(receipt, context(scopeA, {artifactSha256: "0".repeat(64)}));
  assert.deepEqual(intrinsicDigestMismatch, {state: "invalid", deleteProtectedArtifact: true});
  const otherKey = generateKeyPairSync("ed25519").publicKey.export({format: "der", type: "spki"}).subarray(-32);
  const wrongPinnedKey = await validateOfflineDownloadAuthorizationReceipt(receipt, {...context(), pinnedIdentity: {...pinnedIdentity, publicKey: otherKey}});
  assert.equal(wrongPinnedKey.state, "invalid");
});

test("replacement is accepted only after full pinned signature and binding validation", async () => {
  const receipt = signedReceipt();
  const replacement = signedReceipt({receiptId: "receipt-2", lastVerifiedAt: "2026-08-31T12:00:01.000Z"});
  const valid = await revalidateOfflineDownloadAuthorization({
    async revalidateOfflineDownloadAuthorization() { return {outcome: "valid-replacement", receipt: replacement}; }
  }, receipt, context());
  assert.deepEqual(valid, {action: "replace", receipt: replacement});

  const hostile = [
    {...replacement, signature: `${replacement.signature.slice(0, -1)}${replacement.signature.endsWith("A") ? "Q" : "A"}`},
    signedReceipt({receiptId: "receipt-wrong-artifact", artifactOverride: {sha256: "0".repeat(64), sizeBytes: 15}}),
    signedReceipt({receiptId: "receipt-wrong-scope", viewerScope: {...scopeA, profileId: "profile-c"}}),
    signedReceipt({receiptId: "receipt-expired", lastVerifiedAt: "2026-06-01T00:00:00.000Z"})
  ];
  for (const candidate of hostile) {
    const decision = await revalidateOfflineDownloadAuthorization({
      async revalidateOfflineDownloadAuthorization() { return {outcome: "valid-replacement", receipt: candidate}; }
    }, receipt, context());
    assert.deepEqual(decision, {action: "preserve", outcome: "transport-failure"});
  }
  assert.deepEqual(await revalidateOfflineDownloadAuthorization({
    async revalidateOfflineDownloadAuthorization() { throw new Error("offline"); }
  }, receipt, context()), {action: "preserve", outcome: "transport-failure"});
  assert.deepEqual(await revalidateOfflineDownloadAuthorization({
    async revalidateOfflineDownloadAuthorization() { return {outcome: "revoked"}; }
  }, receipt, context()), {action: "delete", outcome: "revoked"});
  assert.deepEqual(await revalidateOfflineDownloadAuthorization({
    async revalidateOfflineDownloadAuthorization() { return {outcome: "out-of-scope"}; }
  }, receipt, context()), {action: "preserve", outcome: "out-of-scope"});
});

File: internal/app/offline_download_authorization_test.go
package app

import (
	"bytes"
	"context"
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"net/http"
	"net/url"
	"strings"
	"testing"
	"time"

	"github.com/PorticoMediaServer/portico-server/internal/config"
)

func TestOfflineDownloadAuthorizationCanonicalizationIsRFC8785AndSignatureIsExact(t *testing.T) {
	receipt := OfflineDownloadAuthorizationReceipt{
		Version: 1, Purpose: offlineDownloadAuthorizationPurpose, ReceiptID: "receipt-é-😀",
		ViewerScope: OfflineDownloadAuthorizationViewerScope{
			ScopeKind: "server-bound", Authority: "local", AccountID: "account", ProfileID: "profile",
			ServerID: "server", AuthorizationRevision: "revision",
		},
		Issuer: OfflineDownloadAuthorizationIssuer{ServerID: "server", SigningKeyFingerprint: "sha256:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"},
		Preparation: OfflineDownloadAuthorizationPreparation{
			PreparationID: "preparation", MediaID: "media", MediaVersionID: "version", QualityID: "source",
		},
		Artifact: OfflineDownloadAuthorizationArtifact{
			SHA256: "2d711642b726b04401627ca9fbac32f5c8530fb1903cc4db02258717921a4881", SizeBytes: 1,
		},
		LastVerifiedAt: "2026-08-30T12:00:00Z", VerifyBy: "2026-09-29T12:00:00Z", Signature: strings.Repeat("ignored", 10),
	}
	payload, err := canonicalOfflineDownloadAuthorizationPayload(receipt)
	if err != nil {
		t.Fatal(err)
	}
	want := `{"artifact":{"sha256":"2d711642b726b04401627ca9fbac32f5c8530fb1903cc4db02258717921a4881","sizeBytes":1},"issuer":{"serverId":"server","signingKeyFingerprint":"sha256:AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"},"lastVerifiedAt":"2026-08-30T12:00:00Z","preparation":{"mediaId":"media","mediaVersionId":"version","preparationId":"preparation","qualityId":"source"},"purpose":"offline-download-authorization","receiptId":"receipt-é-😀","verifyBy":"2026-09-29T12:00:00Z","version":1,"viewerScope":{"accountId":"account","authority":"local","authorizationRevision":"revision","profileId":"profile","scopeKind":"server-bound","serverId":"server"}}`
	if string(payload) != want {
		t.Fatalf("canonical payload\n got: %s\nwant: %s", payload, want)
	}
	seed := sha256.Sum256([]byte("offline-receipt-test-key"))
	privateKey := ed25519.NewKeyFromSeed(seed[:])
	identity := serverIdentity{PrivateKey: privateKey, PublicKey: privateKey.Public().(ed25519.PublicKey)}
	identity.Fingerprint = publicKeyFingerprint(identity.PublicKey)
	record := downloadPreparationRecord{
		DownloadPreparation:   DownloadPreparation{ID: "preparation", MediaID: "media", QualityProfile: "source", SizeBytes: 1},
		AuthorizationRevision: "revision", MediaVersionID: "version", VersionFingerprint: "fence",
		ArtifactSHA256: receipt.Artifact.SHA256,
	}
	signed, err := signOfflineDownloadAuthorizationReceipt(record, receipt.ViewerScope, identity, time.Date(2026, 8, 30, 12, 0, 0, 123, time.UTC))
	if err != nil {
		t.Fatal(err)
	}
	signature, err := decodeCanonicalBase64URL(signed.Signature, ed25519.SignatureSize)
	if err != nil {
		t.Fatal(err)
	}
	signedPayload, err := canonicalOfflineDownloadAuthorizationPayload(signed)
	if err != nil || !ed25519.Verify(identity.PublicKey, signedPayload, signature) {
		t.Fatalf("signed JCS payload did not verify: err=%v", err)
	}
	tampered := signed
	tampered.Artifact.SizeBytes++
	tamperedPayload, _ := canonicalOfflineDownloadAuthorizationPayload(tampered)
	if ed25519.Verify(identity.PublicKey, tamperedPayload, signature) {
		t.Fatal("tampered artifact verified")
	}
}

func TestOfflineReceiptViewerIdentityEnvelopeSeparatesMalformedFromForeign(t *testing.T) {
	scope := OfflineDownloadAuthorizationViewerScope{
		ScopeKind: "server-bound", Authority: "local", AccountID: "account", ProfileID: "profile",
		ServerID: "server", AuthorizationRevision: "revision",
	}
	matching := `{"viewerScope":{"scopeKind":"server-bound","authority":"local","accountId":"account","profileId":"profile","serverId":"server","authorizationRevision":"old-revision"},"issuer":{"serverId":"server"}}`
	foreign := `{"viewerScope":{"scopeKind":"server-bound","authority":"local","accountId":"account","profileId":"other-profile","serverId":"server","authorizationRevision":"old-revision"},"issuer":{"serverId":"server"}}`
	missingIssuer := `{"viewerScope":{"scopeKind":"server-bound","authority":"local","accountId":"account","profileId":"profile","serverId":"server","authorizationRevision":"old-revision"}}`
	tests := []struct {
		name string
		raw  string
		want offlineReceiptEnvelopeDisposition
	}{
		{name: "matching identity ignores stale revision", raw: matching, want: offlineReceiptEnvelopeMatching},
		{name: "complete foreign identity", raw: foreign, want: offlineReceiptEnvelopeForeign},
		{name: "empty object", raw: `{}`, want: offlineReceiptEnvelopeInvalid},
		{name: "missing issuer", raw: missingIssuer, want: offlineReceiptEnvelopeInvalid},
		{name: "malformed JSON", raw: `{"viewerScope":`, want: offlineReceiptEnvelopeInvalid},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			if got := offlineReceiptViewerIdentityEnvelope(json.RawMessage(test.raw), scope); got != test.want {
				t.Fatalf("disposition=%d want=%d", got, test.want)
			}
		})
	}
}

func TestDownloadArtifactDigestWriterCoversOnlyCommittedProducerBytes(t *testing.T) {
	var output bytes.Buffer
	writer := newOptimizedArtifactDigestWriter(&output)
	if _, err := writer.Write([]byte("producer-")); err != nil {
		t.Fatal(err)
	}
	if _, err := writer.Write([]byte("bytes")); err != nil {
		t.Fatal(err)
	}
	digest, size := writer.Digest()
	want := sha256.Sum256([]byte("producer-bytes"))
	if digest != hex.EncodeToString(want[:]) || size != int64(output.Len()) {
		t.Fatalf("producer digest=%q size=%d body=%q", digest, size, output.String())
	}
}

func TestOptimizeCompletionQueuesVerificationWithoutViewerPoll(t *testing.T) {
	_, db, server := newDiscoveryTestServer(t, config.Config{})
	seedDownloadGrantMedia(t, db)
	user, err := server.getUser(adminUserID(t, db))
	if err != nil {
		t.Fatal(err)
	}
	preparation, err := server.createDownloadPreparationContext(context.Background(), user, DownloadPreparationCreateRequest{MediaID: "grant_media", QualityProfile: "source"})
	if err != nil {
		t.Fatal(err)
	}
	record, err := server.downloadPreparationRecordContext(context.Background(), `WHERE dp.id = ? LIMIT 1`, preparation.ID)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := db.Exec(`DELETE FROM jobs WHERE id = ?`, record.JobID); err != nil {
		t.Fatal(err)
	}
	now := time.Now().UTC().Format(time.RFC3339Nano)
	job := Job{ID: "job_optimize_handoff", Type: "optimize_version", ResourceType: "media", ResourceID: record.MediaID}
	if _, err := db.Exec(`
		INSERT INTO jobs (id, type, status, progress, message, resource_type, resource_id, metadata_json,
			leased_by, lease_expires_at, cancellation_requested_at, created_at, updated_at)
		VALUES (?, 'optimize_version', 'running', 92, 'Stored', 'media', ?, '{"profile":"universal-720p"}', ?, ?, '', ?, ?)`,
		job.ID, record.MediaID, server.jobLeaseOwner(job.ID), time.Now().UTC().Add(time.Hour).Format(time.RFC3339Nano), now, now); err != nil {
		t.Fatal(err)
	}
	if _, err := db.Exec(`UPDATE download_preparations SET job_id = ?, state = 'running' WHERE id = ?`, job.ID, record.ID); err != nil {
		t.Fatal(err)
	}
	if err := server.completeOptimizeVersionAndQueueDownloadVerifications(job, "Optimized version complete."); err != nil {
		t.Fatal(err)
	}
	var preparationJobID, preparationState, producerStatus, nextType string
	if err := db.QueryRow(`SELECT job_id, state FROM download_preparations WHERE id = ?`, record.ID).Scan(&preparationJobID, &preparationState); err != nil {
		t.Fatal(err)
	}
	if err := db.QueryRow(`SELECT status FROM jobs WHERE id = ?`, job.ID).Scan(&producerStatus); err != nil {
		t.Fatal(err)
	}
	if err := db.QueryRow(`SELECT type FROM jobs WHERE id = ?`, preparationJobID).Scan(&nextType); err != nil {
		t.Fatal(err)
	}
	if producerStatus != "complete" || preparationState != "queued" || nextType != downloadArtifactVerificationJobType {
		t.Fatalf("producer=%q preparation=%q next=%q", producerStatus, preparationState, nextType)
	}
}

func TestOptimizeFailureAtomicallyFailsLinkedPreparationsWithoutCrossingCancellation(t *testing.T) {
	_, db, server := newDiscoveryTestServer(t, config.Config{})
	seedDownloadGrantMedia(t, db)
	user, err := server.getUser(adminUserID(t, db))
	if err != nil {
		t.Fatal(err)
	}
	preparation, err := server.createDownloadPreparationContext(context.Background(), user, DownloadPreparationCreateRequest{MediaID: "grant_media", QualityProfile: "source"})
	if err != nil {
		t.Fatal(err)
	}
	record, err := server.downloadPreparationRecordContext(context.Background(), `WHERE dp.id = ? LIMIT 1`, preparation.ID)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := db.Exec(`DELETE FROM jobs WHERE id = ?`, record.JobID); err != nil {
		t.Fatal(err)
	}
	now := time.Now().UTC().Format(time.RFC3339Nano)
	job := Job{ID: "job_optimize_failure", Type: "optimize_version", ResourceType: "media", ResourceID: record.MediaID}
	if _, err := db.Exec(`INSERT INTO jobs
		(id,type,status,progress,message,resource_type,resource_id,leased_by,lease_expires_at,cancellation_requested_at,created_at,updated_at)
		VALUES (?, 'optimize_version', 'running', 50, 'running', 'media', ?, ?, ?, '', ?, ?)`,
		job.ID, record.MediaID, server.jobLeaseOwner(job.ID), time.Now().UTC().Add(time.Hour).Format(time.RFC3339Nano), now, now); err != nil {
		t.Fatal(err)
	}
	if _, err := db.Exec(`UPDATE download_preparations SET job_id=?, state='running', artifact_sha256=?, size_bytes=1, size_kind='exact' WHERE id=?`,
		job.ID, strings.Repeat("a", 64), record.ID); err != nil {
		t.Fatal(err)
	}
	if err := server.failOptimizeVersionAndDownloadPreparations(job, "producer failed"); err != nil {
		t.Fatal(err)
	}
	var preparationState, preparationCode, digest, sizeKind, jobStatus, jobCode string
	var size int64
	if err := db.QueryRow(`SELECT state,error_code,artifact_sha256,size_bytes,size_kind FROM download_preparations WHERE id=?`, record.ID).
		Scan(&preparationState, &preparationCode, &digest, &size, &sizeKind); err != nil {
		t.Fatal(err)
	}
	if err := db.QueryRow(`SELECT status,error_code FROM jobs WHERE id=?`, job.ID).Scan(&jobStatus, &jobCode); err != nil {
		t.Fatal(err)
	}
	if preparationState != "failed" || preparationCode != "preparation_failed" || digest != "" || size != 0 || sizeKind != "unknown" || jobStatus != "failed" || jobCode != "optimize_failed" {
		t.Fatalf("preparation=%q/%q digest=%q size=%d/%q job=%q/%q", preparationState, preparationCode, digest, size, sizeKind, jobStatus, jobCode)
	}

	if _, err := db.Exec(`UPDATE jobs SET status='running', leased_by=?, cancellation_requested_at=?, error_code='' WHERE id=?`,
		server.jobLeaseOwner(job.ID), now, job.ID); err != nil {
		t.Fatal(err)
	}
	if _, err := db.Exec(`UPDATE download_preparations SET state='running', error_code='' WHERE id=?`, record.ID); err != nil {
		t.Fatal(err)
	}
	if err := server.failOptimizeVersionAndDownloadPreparations(job, "must not win cancellation"); err != nil {
		t.Fatal(err)
	}
	if err := db.QueryRow(`SELECT state FROM download_preparations WHERE id=?`, record.ID).Scan(&preparationState); err != nil || preparationState != "running" {
		t.Fatalf("cancelled producer overwrote preparation state=%q err=%v", preparationState, err)
	}
}

func TestDownloadPreparationProjectionCannotOverwriteConcurrentLifecycleState(t *testing.T) {
	_, db, server := newDiscoveryTestServer(t, config.Config{})
	seedDownloadGrantMedia(t, db)
	user, err := server.getUser(adminUserID(t, db))
	if err != nil {
		t.Fatal(err)
	}
	preparation, err := server.createDownloadPreparationContext(context.Background(), user, DownloadPreparationCreateRequest{MediaID: "grant_media", QualityProfile: "source"})
	if err != nil {
		t.Fatal(err)
	}
	stale, err := server.downloadPreparationRecordContext(context.Background(), `WHERE dp.id = ? LIMIT 1`, preparation.ID)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := db.Exec(`UPDATE download_preparations SET state = 'cancelled', cancelled_at = ?, updated_at = ? WHERE id = ?`,
		time.Now().UTC().Format(time.RFC3339Nano), time.Now().UTC().Format(time.RFC3339Nano), stale.ID); err != nil {
		t.Fatal(err)
	}
	if _, err := server.materializeDownloadPreparationContext(context.Background(), user, stale); err != nil {
		t.Fatal(err)
	}
	var state string
	if err := db.QueryRow(`SELECT state FROM download_preparations WHERE id = ?`, stale.ID).Scan(&state); err != nil {
		t.Fatal(err)
	}
	if state != "cancelled" {
		t.Fatalf("stale read overwrote authoritative lifecycle state with %q", state)
	}
}

func TestOfflineAuthorizationGrantAndRevalidationAreSecurityFenceAtomic(t *testing.T) {
	serverURL, db, server := newDiscoveryTestServer(t, config.Config{})
	client := authenticatedDownloadGrantClient(t, serverURL)
	seedDownloadGrantMedia(t, db)
	var preparation downloadPreparationView
	status, body := doJSON(t, client, http.MethodPost, serverURL+"/api/download-preparations", DownloadPreparationCreateRequest{MediaID: "grant_media", QualityProfile: "source"}, &preparation)
	if status != http.StatusCreated {
		t.Fatalf("create preparation status=%d body=%s", status, body)
	}
	preparation = completeDownloadPreparationForTest(t, server, client, serverURL, preparation)
	userID := adminUserID(t, db)
	var originalPermissionsJSON string
	if err := db.QueryRow(`SELECT permissions_json FROM users WHERE id = ?`, userID).Scan(&originalPermissionsJSON); err != nil {
		t.Fatal(err)
	}
	revoke := func() {
		if _, err := db.Exec(`UPDATE users SET permissions_json = json_set(permissions_json, '$.downloadMedia', json('false')) WHERE id = ?`, userID); err != nil {
			panic(err)
		}
		server.invalidateAuthorizationCachesForMutation("UPDATE users SET permissions_json", []string{"users"})
	}
	restore := func() {
		if _, err := db.Exec(`UPDATE users SET permissions_json = ? WHERE id = ?`, originalPermissionsJSON, userID); err != nil {
			t.Fatal(err)
		}
		server.invalidateAuthorizationCachesForMutation("UPDATE users SET permissions_json", []string{"users"})
	}
	server.offlineAuthorizationBeforeCommitHook = revoke
	status, body = doJSON(t, client, http.MethodPost, serverURL+"/api/download-preparations/"+url.PathEscape(preparation.ID)+"/grant", downloadPreparationGrantRequest{Delivery: "native"}, nil)
	server.offlineAuthorizationBeforeCommitHook = nil
	if status != http.StatusConflict || !strings.Contains(body, "download_preparation_not_ready") {
		t.Fatalf("revoked-before-commit grant status=%d body=%s", status, body)
	}
	var grants int
	if err := db.QueryRow(`SELECT COUNT(*) FROM media_download_grants WHERE preparation_id = ?`, preparation.ID).Scan(&grants); err != nil || grants != 0 {
		t.Fatalf("stale authority published %d grants: %v", grants, err)
	}
	restore()
	var grant MediaDownloadGrantResponse
	status, body = doJSON(t, client, http.MethodPost, serverURL+"/api/download-preparations/"+url.PathEscape(preparation.ID)+"/grant", downloadPreparationGrantRequest{Delivery: "native"}, &grant)
	if status != http.StatusCreated || grant.AuthorizationReceipt == nil {
		t.Fatalf("native grant status=%d body=%s grant=%#v", status, body, grant)
	}
	receipt := *grant.AuthorizationReceipt
	rawReceipt, _ := json.Marshal(receipt)
	server.offlineAuthorizationBeforeCommitHook = revoke
	var revoked OfflineDownloadAuthorizationRevalidationResponse
	status, body = doJSON(t, client, http.MethodPost, serverURL+"/api/offline-download-authorizations/revalidate",
		OfflineDownloadAuthorizationRevalidationRequest{Receipt: rawReceipt}, &revoked)
	server.offlineAuthorizationBeforeCommitHook = nil
	if status != http.StatusOK || revoked.Outcome != "revoked" || revoked.Receipt != nil {
		t.Fatalf("revocation race status=%d body=%s response=%#v", status, body, revoked)
	}
	restore()
	var replacement OfflineDownloadAuthorizationRevalidationResponse
	status, body = doJSON(t, client, http.MethodPost, serverURL+"/api/offline-download-authorizations/revalidate",
		OfflineDownloadAuthorizationRevalidationRequest{Receipt: rawReceipt}, &replacement)
	if status != http.StatusOK || replacement.Outcome != "valid-replacement" || replacement.Receipt == nil || replacement.Receipt.ReceiptID == receipt.ReceiptID {
		t.Fatalf("replacement status=%d body=%s response=%#v", status, body, replacement)
	}
	if !offlineReceiptMatchesPreparation(*replacement.Receipt, downloadPreparationRecord{
		DownloadPreparation: DownloadPreparation{ID: receipt.Preparation.PreparationID, MediaID: receipt.Preparation.MediaID, QualityProfile: receipt.Preparation.QualityID, SizeBytes: receipt.Artifact.SizeBytes},
		MediaVersionID:      receipt.Preparation.MediaVersionID, ArtifactSHA256: receipt.Artifact.SHA256,
	}) {
		t.Fatal("replacement changed immutable preparation or artifact binding")
	}
	tampered := receipt
	tampered.ViewerScope.ProfileID = "another-profile"
	tamperedRaw, _ := json.Marshal(tampered)
	var outOfScope OfflineDownloadAuthorizationRevalidationResponse
	status, body = doJSON(t, client, http.MethodPost, serverURL+"/api/offline-download-authorizations/revalidate",
		OfflineDownloadAuthorizationRevalidationRequest{Receipt: tamperedRaw}, &outOfScope)
	if status != http.StatusOK || outOfScope.Outcome != "out-of-scope" || strings.Contains(body, preparation.ID) {
		t.Fatalf("cross-profile receipt status=%d body=%s response=%#v", status, body, outOfScope)
	}
	tampered = receipt
	tampered.Artifact.SHA256 = strings.Repeat("0", 64)
	tamperedRaw, _ = json.Marshal(tampered)
	var invalid OfflineDownloadAuthorizationRevalidationResponse
	status, body = doJSON(t, client, http.MethodPost, serverURL+"/api/offline-download-authorizations/revalidate",
		OfflineDownloadAuthorizationRevalidationRequest{Receipt: tamperedRaw}, &invalid)
	if status != http.StatusOK || invalid.Outcome != "invalid" {
		t.Fatalf("tampered receipt status=%d body=%s response=%#v", status, body, invalid)
	}
}
